src/neural/cuda/layers.cc

```diff
@@ -1225,6 +1141,425 @@ ResidualBlock<DataType>::~ResidualBlock() {
+  printf("\nsize of ln2 betas/gammas: %d, %d\n",
+         (int)cpu_weights.ln2_betas.size(), (int)cpu_weights.ln2_gammas.size());
+  printf("\nsize of weight ffn.dense1_b/w: %d, %d\n",
+         (int)cpu_weights.ffn.dense1_b.size(),
+         (int)cpu_weights.ffn.dense1_w.size());
+  printf("\nsize of weight ffn.dense2_b/w: %d, %d\n",
+         (int)cpu_weights.ffn.dense2_b.size(),
+         (int)cpu_weights.ffn.dense2_w.size());
+
+
+  allocAndUpload<DataType>(&mha_q_w, cpu_weights.mha.q_w, scratch);
+  allocAndUpload<DataType>(&mha_q_b, cpu_weights.mha.q_b, scratch);
+
+  allocAndUpload<DataType>(&mha_k_w, cpu_weights.mha.k_w, scratch);
+  allocAndUpload<DataType>(&mha_k_b, cpu_weights.mha.k_b, scratch);
+
+  allocAndUpload<DataType>(&mha_v_w, cpu_weights.mha.v_w, scratch);
+  allocAndUpload<DataType>(&mha_v_b, cpu_weights.mha.v_b, scratch);
+
+  allocAndUpload<DataType>(&mha_dense_w, cpu_weights.mha.dense_w, scratch);
+  allocAndUpload<DataType>(&mha_dense_b, cpu_weights.mha.dense_b, scratch);
+
+
+  allocAndUpload<DataType>(&ln1_gammas, cpu_weights.ln1_gammas, scratch);
+  allocAndUpload<DataType>(&ln1_betas, cpu_weights.ln1_betas, scratch);
+
+  allocAndUpload<DataType>(&ffn_dense1_w, cpu_weights.ffn.dense1_w, scratch);
+  allocAndUpload<DataType>(&ffn_dense1_b, cpu_weights.ffn.dense1_b, scratch);
+
+  allocAndUpload<DataType>(&ffn_dense2_w, cpu_weights.ffn.dense2_w, scratch);
+  allocAndUpload<DataType>(&ffn_dense2_b, cpu_weights.ffn.dense2_b, scratch);
+
+  allocAndUpload<DataType>(&ln2_gammas, cpu_weights.ln2_gammas, scratch);
+  allocAndUpload<DataType>(&ln2_betas, cpu_weights.ln2_betas, scratch);
+}
+
+// taken from https://stackoverflow.com/questions/1659440/32-bit-to-16-bit-floating-point-conversion
+unsigned short float_to_half(const float x) {
+  const unsigned int b = (*(unsigned int*)&x) + 0x00001000;
+  const unsigned int e = (b & 0x7F800000) >> 23;
+  const unsigned int m = b & 0x007FFFFF;
+  return (b & 0x80000000) >> 16 |
+         (e > 112) * ((((e - 112) << 10) & 0x7C00) | m >> 13) |
+         ((e < 113) & (e > 101)) *
+             ((((0x007FF000 + m) >> (125 - e)) + 1) >> 1) |
+         (e > 143) * 0x7FFF;
+}
+
+
+template <>
+void AttentionPolicyHead<half>::Eval(
+    int N, half* output, const half* input, const half* input2,
+    void* scratch, size_t scratch_size, cudnnHandle_t cudnn,
+    cublasHandle_t cublas, cudaStream_t stream) {
+
+  half* scratch0 = (half*)scratch;
+  half* scratch1 = (half*)scratch + scratch_size / (2 * sizeof(half));
+  half* scratch2 = (half*)input2;
+  half* scratch3 = (half*)input2 + scratch_size / (2 * sizeof(half));
+  half* scratch4 = output + scratch_size / (2 * sizeof(half));
+
+  // half alpha = float2half_rn(1.0f), beta = float2half_rn(0.0f);
+  const __half_raw one_h{0x3C00};
+  const __half_raw zero_h{0};
+  half alpha = one_h;
+  half beta = zero_h;
+
+
+  fp16NCHWtoNHWC(scratch1, input, N, 64, N, 64, 8, 8);
```

Comment:
The 64's in this line should be input_->GetC()

Reply:
thanks. Fixed.